src/librustc/mir/transform/dataflow.rs

```diff
@@ -0,0 +1,419 @@
+use mir::repr as mir;
+use mir::cfg::CFG;
+use mir::repr::BasicBlock;
+use rustc_data_structures::bitvec::BitVector;
+
+use mir::transform::lattice::Lattice;
+
+
+pub trait DataflowPass<'tcx> {
+    type Lattice: Lattice;
+    type Rewrite: Rewrite<'tcx, Self::Lattice>;
+    type Transfer: Transfer<'tcx, Self::Lattice>;
+}
+
+pub trait Rewrite<'tcx, L: Lattice> {
+    /// The rewrite function which given a statement optionally produces an alternative graph to be
+    /// placed in place of the original statement.
+    ///
+    /// The 2nd BasicBlock *MUST NOT* have the terminator set.
+    ///
+    /// Correctness precondition:
+    /// * transfer_stmt(statement, fact) == transfer_stmt(rewrite_stmt(statement, fact))
+    /// that is, given some fact `fact` true before both the statement and relacement graph, and
+    /// a fact `fact2` which is true after the statement, the same `fact2` must be true after the
+    /// replacement graph too.
+    fn stmt(&mir::Statement<'tcx>, &L, &mut CFG<'tcx>) -> StatementChange<'tcx>;
```

Comment:
It might be nicer for these methods to take `&self`. This would allow runtime customization of what passes do. Additionally, it might make specifying compound rewrites more ergonomic by creating an `and_then` method on rewrite creating a `RewriteAndThen`. Thus,

```rust
impl<'tcx> DataflowPass<'tcx> for ACSPropagate {
    type Lattice = ACSLattice<'tcx>;
    type Rewrite = RewriteAndThen<'tcx, AliasRewrite,
                                  RewriteAndThen<'tcx, ConstRewrite, SimplifyRewrite>>;
    type Transfer = ACSPropagateTransfer;
}

ar_forward::<ACSPropagateTransfer, ACSPropagate>(&mut mir.cfg, Facts::new(), q);
```

would turn into

```rust
ar_forward(
    &mut mir.cfg,
    Facts::new(), q,
    ACSPropogateTransfer,
    AliasRewrite.and_then(ConstRewrite).and_then(SimplifyRewrite)
);
```

This would require that `ar_forward` take its rewrite function and analysis by value instead of by type.